lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+/** @defgroup fdcan_file FDCAN peripheral API
+ * @ingroup peripheral_apis
+ * 
+ * @brief <b>libopencm3 STM32G4xx CAN</b>
+ * 
+ * @version 1.0.0
+ * 
+ * @author @htmlonly &copy; @endhtmlonly 2021 Eduard Drusa <ventyl86 at netkosice dot sk>
+ * 
+ * Devices can have up to three FDCAN peripherals residing in one FDCAN block. The peripherals 
+ * support both CAN 2.0 A and B standard and Bosch FDCAN standard. FDCAN frame format and
+ * bitrate switching is supported. The peripheral has several filters for incoming messages that
+ * can be distributed between two FIFOs and three transmit mailboxes. For transmitted messages
+ * it is possible to opt for event notification once message is transmitted.
+ * 
+ * LGPL License Terms @ref lgpl_license
+*/
+
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2021 Eduard Drusa <ventyl86 at netkosice dot sk>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <libopencm3/stm32/fdcan.h>
+#include <libopencm3/stm32/rcc.h>
+
+
+/* Timeout for FDCAN_CCCR register INIT bit to accept set value.
+ * This timeout is required because FDCAN uses two different clocks
+ * feeding two different portions of block. There can be slight delay
+ * based on how clocks are set up. While amount of FDCAN_clk /
+ * FDCAN_pclk combinations is high and clock speeds may vary a lot,
+ * following value has been choosen as sane default.
+ */
+#define FDCAN_CCCR_INIT_TIMEOUT		0x0000FFFF
+
+static int fdcan_cccr_init_cfg(uint32_t canport, bool set);
+static int fdcan_get_fill_rxfifo(uint32_t canport, uint8_t fifo_id, 
+		unsigned * get_index, unsigned * pending_frames);
+static struct FDCAN_MessageRAM * fdcan_get_msgram_addr(uint32_t canport);
+static int fdcan_get_free_txbuf(uint32_t canport);
+static uint32_t fdcan_length_to_dlc(uint8_t length);
+static uint8_t fdcan_dlc_to_length(uint32_t dlc);
+
+/** @ingroup fdcan_file */
+/**@{
+ * */
+
+/** @brief FDCAN Reset
+ *
+ * This will reset the FDCAN peripheral. All registers will be placed in
+ * the reset state. Note that it is not possible to reset FDCAN per-CAN
+ */
+void fdcan_reset(void)
+{
+	rcc_periph_reset_pulse(RST_FDCAN);
+}
```

Comment:
I would suggest that this type of api duplication not be encouraged, just drop the function entirely. peripheral reset is already handled (twice) in the rcc apis.

Reply:
Will do